shared/middleware/helpers/bucket_helper.rb

```diff
@@ -175,43 +175,80 @@ def create_or_replace(encrypted_channel_id, filename, body, version = nil, abuse
     response
   end
 
-  def check_current_version(encrypted_channel_id, filename, version_to_replace, timestamp, tab_id, user_id)
-    return unless filename == 'main.json' && @bucket == CDO.sources_s3_bucket && version_to_replace
+  # When updating s3://cdo-v3-sources/.../main.json, checks that the
+  # current_version from the client is the latest version on the server. If a
+  # different client more recently wrote to this project, logs an event to
+  # firehose and halts with 409 Conflict.
+  #
+  # In some cases, S3 replication lag could cause the current_version not to
+  # even appear in the version list. In this case, do not log or halt.
+  #
+  # Clients displaying projects must obey the following rules:
+  # (1) When loading a project, read its latest version.
+  # (2) When saving a project for the first time, create a new version, but
+  #     do not replace any previous versions.
+  # (3) When saving a project subsequent times, create a new version, and
+  #     replace the "current version" (i.e. last known version that client wrote).
+  #
+  # Therefore, a project will only ever replace a version that it created, and
+  # we can say the client "owns" a particular version if that client created it.
```

Comment:
:tada: This is a great explanation, and a very well-reasoned approach to this problem.